lang/core/os/is_virtual.go

```diff
@@ -0,0 +1,160 @@
+// license, the licensors of this program grant you additional permission to
+// convey the resulting work. Furthermore, the licensors of this program grant
+// the original author, James Shubin, additional permission to update this
+// additional permission if he deems it necessary to achieve the goals of this
+// additional permission.
+
+package coreos
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/purpleidea/mgmt/lang/funcs/simple"
+	"github.com/purpleidea/mgmt/lang/types"
+	"github.com/purpleidea/mgmt/util/errwrap"
+)
+
+// The Desktop Management Interface (DMI) is a framwork for exposing system data
+// to software. We will make use of it to determine whether we're using a
+// virtualization platform. dmiFilesSlice is a slice of file paths which might
+// contain information related to the presence of a virtualization platform.
+// This is where we might find the values present in virtualizationVendorSlice.
+var dmiFilesSlice = []string{
+	"/sys/class/dmi/id/product_name",
+	"/sys/class/dmi/id/sys_vendor",
+	"/sys/class/dmi/id/board_vendor",
+	"/sys/class/dmi/id/bios_vendor",
+	"/sys/class/dmi/id/product_version",
+}
+
+// virtualizationVendorSlice is a slice of strings that might be found on DMI
+// related files (dmiFilesSlice) during the checks performed for the presence of
+// virtualization.
+var virtualizationVendorSlice = []string{
+	"KVM",
+	"OpenStack",
+	"KubeVirt",
+	"Amazon EC2",
+	"QEMU",
+	"VMware",
+	"VMW",
+	"innotek GmbH",
+	"VirtualBox",
+	"Oracle Corporation",
+	"Xen",
+	"Bochs",
+	"Parallels",
+	"BHYVE",
+	"Hyper-V",
+	"Apple Virtualization",
+	"Google Computer Engine",
+}
+
+func init() {
+	simple.ModuleRegister(ModuleName, "is_virtual", &simple.Scaffold{
+		T: types.NewType("func() bool"),
+		F: IsVirtual,
+	})
+}
+
+// IsVirtual is a shim for the isVirtual function,
+func IsVirtual(ctx context.Context, input []types.Value) (types.Value, error) {
+	b, err := isVirtual(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &types.BoolValue{V: b}, nil
+}
+
+// isVirtual is a function that executes two types of checks: first, it checks
+// whether we're running on Linux. If that's the case, we run checks related
+// with the presence of virtualization platforms. If any of those checks returns
+// true, then so does this function. Otherwise, it's assumed that it's not a
+// virtualized environment.
+func isVirtual(ctx context.Context) (bool, error) {
+	// if we implement detection for OS other than Linux, this logic will have
+	// to change
+	opersys := runtime.GOOS
+	if opersys != "linux" {
+		return false, errors.New("operating system is not Linux")
+	}
+
+	cpuInfoCheck, err1 := cpuInfoCheck(ctx)
+	dmiFileCheck, err2 := dmiFileCheck(ctx)
+
+	switch {
+	case cpuInfoCheck:
+		return true, nil
+	case dmiFileCheck:
+		return true, nil
+	case err1 != nil || err2 != nil:
+		return false, errwrap.Append(err1, err2)
```

Comment:
Did you check what `errwrap.Append(nil, err2)` or `errwrap.Append(err1, nil)` returns?

I think the logic of this function can be simplified too.

What do you think of:

cpuInfoCheck, err1 := cpuInfoCheck(ctx)
if err1 == nil && cpuInfoCheck {
    return true, nil
}

dmiFileCheck, err2 := dmiFileCheck(ctx)
if err2 == nil && dmiFileCheck {
    return true, nil
}

if err1 == nil && err2 == nil {
    return false, nil
}

return ...

What do you think?

My unwritten rule is you don't want to use switch unless you really have a lot of options.